POEM_003.md

```diff
@@ -0,0 +1,111 @@
+POEM ID: 003  
+Title: Adding a pre_setup method to create objects and communicators  
+Author: anilyil (Anil Yildirim)  
+Competing POEMs: N/A  
+Related POEMs: N/A  
+Associated implementation PR: N/A  
+
+Status: 
+
+ - [x] Active
+ - [ ] Requesting decision
+ - [ ] Accepted
+ - [ ] Rejected
+
+
+Problem statement
+=================
+
+OpenMDAO problems contain a method called `setup`, which is called by the user after the user specifies all components and connections.
+No fundamental changes to the components can be done after this method is called; in particular, sizes of the intputs and outputs for any component cannot be changed.
+Furthermore, the communicators are not created until `setup` is called, therefore users cannot initialize supplementary Python objects that require access to the communicator.
+
+Frequently, these objects that require the communicator needs to be initialized to determine the input and output sizes for their components, effectively creating a circular dependency.
+This factor makes it difficult to integrate complied analysis codes in OpenMDAO components in a modular way.
+This problem is currently avoided by ad hoc implementations, where the Python objects pass information in a layer hidden from OpenMDAO, which is not ideal and limits the flexibility of the framework.
+
+
+
+
+Suggested solution
+==================
+
+We suggest adding a `pre_setup` method to the `Problem` class in OpenMDAO.
+This method will:
+
+1. Create the communicators,
+2. traverse down the model hierarchy and call the `pre_setup` method on all components, and
+3. traverse up the model hierarchy and call the `pre_configure` method on all components.
+
+After `pre_setup` is called, the user will still be able to make new connections in the model, and define new input and outputs for any component.
```

Comment:
The existing `configure` method already traverse up the model hierarhchy. `setup` traverses down, assigning comms to all group and component instances, and then `configure` is called on the way back up. 

You are allowed to define new i/o for components and new connections for groups each ones `configure` implementation.